Message overhaul

rafaelsamenezes opened the pull request:
This is a big PR, it revamps the message system of ESBMC, removes every cout/cerr call and also add supports to the fmt library. To make the review easier I will go through my main intentions and shortcuts that I've taken during the development.

# Message

The message system is developed inside the `src/util/message`

The idea of the new message system is to have a centralized system that is able to control every output stream into different levels of verbosity. This is done with the `messaget `class which should be passed  as a const reference inside a common ESBMC flow.

Inside a `messaget` there are a list of `message_handlers`, which do the dirty work of pushing into a file, log, terminal input. In future they can be extended to also contain decorators.

Finally, the module also contain fmt specializations for some data structures of esbmc.

# How I've implemented

It is too much work redesign every class to pass down the messaget object. Also, most classes (besides the frontend) only do output to report errors. So the implementation went as:

- `Dump` methods are always for debug and uses the default_message object. If the output is needed for status, then I've called through the fmt::format and use created specializations made for it.
- ~If there is a `std::cerr` followed by an abort/throw, now it will only throw an `std::runtime_error` with the cerr contents on it.~
- If the method needs an ostream, then an ostringstream is used.
- In LLVM dumps, it was used an raw_os_stream with ostringstream
- No includes for iostream are allowed. The only exception is when using `std::cin`. This was isolated into a unique cpp file.
- `std::endl` was replaced with `"\n"`. When at the end of the phrase I either removed the "\n"  (messaget adds a "\n" at each call) or appended it to the end of it.
- Some C API (from the solvers) need an FILE* output, to handle those, it was created an temporary file that sends the output into the messaget.
- I've refactored some methods to make it easier for me.

rafaelsamenezes committed:
added verbosity enum

rafaelsamenezes committed:
removed graphml from ui_message

rafaelsamenezes committed:
removed language uit

rafaelsamenezes committed:
removed ui_message_handler dependencies

rafaelsamenezes committed:
removed ui_message

rafaelsamenezes committed:
added fmt

rafaelsamenezes committed:
code-style fixes

rafaelsamenezes committed:
centralized messaget for all IO operations

rafaelsamenezes committed:
removed unused variable

rafaelsamenezes committed:
revamped message system

rafaelsamenezes committed:
removed unused functions

rafaelsamenezes committed:
renamed std::endl to "\n"

rafaelsamenezes committed:
added bigint support for fmt

rafaelsamenezes committed:
reomoved cout/cerr calls from src/esbmc

rafaelsamenezes committed:
fixed compilation issues on src/esbmc

rafaelsamenezes committed:
fixed werror issues

rafaelsamenezes committed:
removed cout comments

rafaelsamenezes committed:
removed cout as default paremeter

rafaelsamenezes committed:
xml is now in a separate module

rafaelsamenezes committed:
added messaget parameter

rafaelsamenezes committed:
removed irep_debug comments with cout

rafaelsamenezes committed:
fixed compilation issues

rafaelsamenezes committed:
removed iostream

rafaelsamenezes committed:
fixed build for all solvers

rafaelsamenezes committed:
Update .gitignore

rafaelsamenezes committed:
Fixed rebase problem

rafaelsamenezes committed:
Revert "Fixed rebase problem"

This reverts commit d2a9e844e45e8eaf850f2bae0e8fdfac51bdf915.

rafaelsamenezes committed:
fixed rebase problem

rafaelsamenezes committed:
more rebase issues

rafaelsamenezes committed:
added boost libraries dependency

rafaelsamenezes committed:
Removed unused library

rafaelsamenezes committed:
code-style fixes

mikhailramalho commented:
oh, that's nice!

A couple of comments:
1. Could you use msg.error + abort instead of throwing exceptions? Throwing exception adds another layer of complexity when debugging issues.
2. I see a few places that still use std::cerr, are you planning to replace them with msg.error?
3. The reason the dump methods don't have arguments is that debuggers sometimes cannot resolve arguments, so we need to have a simpler method to dump an object's content. Could you revert them to not having an argument?
4. Do we really need a `message_handlert`? I mean, can't `messaget` use fmt internally without the user having to explicitly create the interface?

Minor comment: it always bugged me that we are inconsistent with the `*` in various places. Sometimes we use it for warnings and other times for progress (like the k-induction steps). How about we move them to the `messaget` class? I mean, if it's warning, error or progress, we prepend the message with a few `*` (5 maybe?).

lucasccordeiro commented on src/esbmc/bmc.cpp:
```diff
@@ -226,84 +181,31 @@ smt_convt::resultt bmct::run_decision_procedure(
-  default:
-    assert(false);
-  }
+  msg.status("\nVERIFICATION SUCCESSFUL");
```
Shall we add `\n` after `SUCCESSFUL`?

rafaelsamenezes replied:
Every `status` adds a `\n`. The original statement was "\nVERIFICATION SUCCESSFUL\n"

lucasccordeiro commented on src/esbmc/bmc.cpp:
```diff
@@ -440,16 +340,16 @@ void bmct::report_result(smt_convt::resultt &res)
 
   default:
-    error("decision procedure failed");
+    msg.error("decision procedure failed");
```
Shall we replace `decision procedure failed` by `SMT solver failed`?

lucasccordeiro commented on src/esbmc/bmc.cpp:
```diff
@@ -667,21 +564,19 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
   catch(std::bad_alloc &)
   {
-    std::cout << "Out of memory" << std::endl;
+    msg.error("Out of memory\n");
```
When does this `Out of memory` happen? If we know, we could say in which stage of the verification process it happens (e.g., symbolic execution, SMT solving, etc).

mikhailramalho replied:
That's the problem of having these catch-all exception handling, there is no way of knowing where they came from.

rafaelsamenezes replied:
> When does this Out of memory happen? If we know, we could say in which stage of the verification process it happens (e.g., symbolic execution, SMT solving, etc).

If you want to know specifically where it happened you need to create inner try-catches. For example, during Z3 solving if it catches the `bad_alloc` then you can throw an Z3_bad_alloc (which inherites std::bad_alloc) exception and then try again with a different strategy/solver. Thats the idea of exception handling.

> That's the problem of having these catch-all exception handling, there is no way of knowing where they came from.

The best approach would be to specialize everything that you want to handle differently. However, if you intend to debug the issue it won't make any difference. Most debuggers will stop if an exception is thrown (specially, bad_alloc). Also, if the exception is not handled, the default behaviour is to throw an abort anyway.

lucasccordeiro commented on src/esbmc/bmc.cpp:
```diff
@@ -752,8 +649,9 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
-                  << " the program\n";
+        msg.status(
+          "No VCC remaining, no SMT formula will be generated for"
+          " the program\n");
```
replace `the proram` by `this program`.

lucasccordeiro commented on src/goto-symex/builtin_functions.cpp:
```diff
@@ -467,10 +467,7 @@ void goto_symext::intrinsic_switch_to(
-    std::cerr << "Can't switch to non-constant thread id no";
-    abort();
-  }
+    throw std::runtime_error("Can't switch to non-constant thread id no");
```
`Can't switch to non-constant thread id no` -> `Can't switch to non-constant thread id number`. Shall we output the number?

rafaelsamenezes replied:
We don't know if it is a number. I've changed to output the contents of expr as well

lucasccordeiro reviewed (changes requested):
This is one of the largest PR I have seen in the ESBMC project. Many thanks for improving the ESBMC message output implementation. It is a really nice contribution, which was hanging around for quite a long time.

mikhailramalho commented:
Wow, esbmc is now using all my 32GB of RAM to build this branch D:

mikhailramalho commented:
> Wow, esbmc is now using all my 32GB of RAM to build this branch D:

Reconfiguring cmake seems to fix it. Go figures.

rafaelsamenezes commented:
> Could you use msg.error + abort instead of throwing exceptions? Throwing exception adds another layer of complexity when debugging issues

I know that Fail-Fast is usually better for debugging however exceptions are better if we ever want to have other interfaces such as GUI, plugins or web. 

> I see a few places that still use std::cerr, are you planning to replace them with msg.error?

Yes, I haven't seen them. I will remove those.

> The reason the dump methods don't have arguments is that debuggers sometimes cannot resolve arguments, so we need to have a simpler method to dump an object's content. Could you revert them to not having an argument?

Can I assume that the dump method is intended for the debugger only? If so, I can create always create an `fmt_message_handler` for stdout

> Do we really need a message_handlert? I mean, can't messaget use fmt internally without the user having to explicitly create the interface?

I want to create other message_handlers for different kinds of outputs. However, I could make a default constructor that uses the fmt with stdout/stderr, would that work?

mikhailramalho commented:
Em sex., 2 de jul. de 2021 às 12:11, Rafael Sá Menezes <
***@***.***> escreveu:

> Could you use msg.error + abort instead of throwing exceptions? Throwing
> exception adds another layer of complexity when debugging issues
>
> I know that Fail-Fast is usually better for debugging however exceptions
> are better if we ever want to have other interfaces such as GUI, plugins or
> web.
>

Indeed, but we can also try to support GUI, plugins, or web when/if we have
plans to support them. Right now it's better to preserve the current
behavior.


> I see a few places that still use std::cerr, are you planning to replace
> them with msg.error?
>
> Yes, I haven't seen them. I will remove those.
>
> The reason the dump methods don't have arguments is that debuggers
> sometimes cannot resolve arguments, so we need to have a simpler method to
> dump an object's content. Could you revert them to not having an argument?
>
> Can I assume that the dump method is intended for the debugger only? If
> so, I can create a static messaget that will use the fmt_message_handler
> (for debug builds).
>
It's mostly used via debugger but also when we want to dump the object
along with an error message. I think some of the error messages in the
frontend used to dump the whole object.


> Do we really need a message_handlert? I mean, can't messaget use fmt
> internally without the user having to explicitly create the interface?
>
> I want to create other message_handlers for different kinds of outputs.
> However, I could make a default constructor that uses the fmt with
> stdout/stderr, would that work?
>

Sure, or create an enum to select the output interface when constructing a
messaget object. I'm fine with just simplifying the object creation.

—
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/456#issuecomment-873108064>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH62QQI3C5KYQNCCN7DTVXQJNANCNFSM466JH6UQ>
> .
>


-- 

Mikhail Ramalho.

rafaelsamenezes commented:
>  I think some of the error messages in the frontend used to dump the whole object.

Yeah, I've changed them to be printable through fmt.

> Sure, or create an enum to select the output interface when constructing a messaget object. I'm fine with just simplifying the object creation.

I will make a specialization for the default implementation and replace it. This will solve the dump issue as well.

> Indeed, but we can also try to support GUI, plugins, or web when/if we have plans to support them. Right now it's better to preserve the current behavior.

Ok then

rafaelsamenezes committed:
added default message

rafaelsamenezes committed:
Reverted dump(const messaget&) into dump()

rafaelsamenezes committed:
rolled back exceptions

rafaelsamenezes committed:
removed default_message from aborts

rafaelsamenezes commented:
> Minor comment: it always bugged me that we are inconsistent with the * in various places. Sometimes we use it for warnings and other times for progress (like the k-induction steps). How about we move them to the messaget class? I mean, if it's warning, error or progress, we prepend the message with a few * (5 maybe?).

Sorry, I haven't seen this comment. I like the idea and it will be simple to implement. However, this PR is getting quite big already 3k+ lines. I will open an issue afterwards.

rafaelsamenezes committed:
asserts now throw abort as well

Some compilers are not able to detect that assert(0 && msg) is always false :-1:

rafaelsamenezes committed:
added default_message dependency

rafaelsamenezes committed:
removed last bits of cerr/cout

rafaelsamenezes committed:
code-style fixes

rafaelsamenezes committed:
changed status messages

rafaelsamenezes commented:
@mikhailramalho `debug` reverted back to not having parameter and the exceptions were reverted back to abort

lucasccordeiro reviewed (approved):
LGTM.

mikhailramalho commented on src/clang-c-frontend/clang_c_adjust_expr.cpp:
```diff
@@ -497,17 +500,16 @@ void clang_c_adjust::adjust_type(typet &type)
-      symbol.dump();
-      abort();
+      throw std::runtime_error(
+        fmt::format("expected type symbol, but got\n{}", symbol));
```
Could you change it to message.error?

mikhailramalho commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -35,8 +36,8 @@ clang_c_languaget::clang_c_languaget()
-    std::cerr << "Can't find temporary directory (needed to dump clang headers)"
-              << std::endl;
+    assert(
+      0 && "Can't find temporary directory (needed to dump clang headers)");
```
Could you change it to message.error?

mikhailramalho commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -45,9 +46,8 @@ clang_c_languaget::clang_c_languaget()
-      << "Can't create temporary directory (needed to dump clang headers)"
-      << std::endl;
+    assert(
+      0 && "Can't create temporary directory (needed to dump clang headers)");
```
Could you change it to message.error?

mikhailramalho commented on src/clang-c-frontend/clang_c_language.cpp:
```diff
@@ -74,7 +74,9 @@ void clang_c_languaget::build_compiler_args(const std::string &&tmp_dir)
-    std::cerr << "Unknown word size: " << config.ansi_c.word_size << std::endl;
+    assert(
+      0 &&
+      fmt::format("Unknown word size: {}\n", config.ansi_c.word_size).c_str());
```
Could you change it to message.error?

mikhailramalho commented on src/clang-c-frontend/typecast.cpp:
```diff
@@ -52,6 +52,6 @@ void gen_typecast_to_union(exprt &e, const typet &t)
   msg << "Couldn't map type " << e.type().pretty_name() << " into the union"
-      << std::endl;
+      << "\n";
   throw std::domain_error(msg.str());
```
Could you change it to message.error?

mikhailramalho commented on src/esbmc/bmc.cpp:
```diff
@@ -144,34 +118,18 @@ void bmct::error_trace(
+  oss << "\n"
+      << "Counterexample:"
+      << "\n";
+  show_goto_trace(oss, ns, goto_trace, msg);
```
single line?

mikhailramalho commented on src/esbmc/bmc.cpp:
```diff
@@ -695,11 +591,11 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
 
   if(options.get_bool_option("double-assign-check"))
-    eq->check_for_duplicate_assigns();
+    eq->check_for_duplicate_assigns(msg);
```
doesn't eq already have the message object?

mikhailramalho commented on src/esbmc/esbmc_parseoptions.cpp:
```diff
@@ -80,8 +83,8 @@ struct resultt
-  std::cout << "Timed out" << std::endl;
-
+  default_message msg;
+  msg.error("Timed out");
```
could you change `v` to be a `VerbosityLevel` object? So we can remove the cast.

mikhailramalho commented on src/esbmc/show_vcc.cpp:
```diff
@@ -18,21 +19,10 @@ void bmct::show_vcc(
+  out << "\n"
+      << "VERIFICATION CONDITIONS:"
+      << "\n"
+      << "\n";
```
single string?

mikhailramalho commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -20,15 +20,6 @@ Author: Daniel Kroening, kroening@kroening.com
-#else
-#define DEBUGLOC
-#endif
-
```
This goes back to @lucasccordeiro's doc!

mikhailramalho commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -52,9 +43,8 @@ void goto_convertt::finish_gotos(goto_programt &dest)
-        std::cerr << "goto label " << goto_label << " not found";
-        i.code->dump();
-        abort();
+        throw std::runtime_error(
```
could you replace it with message.error?

mikhailramalho commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -432,8 +421,7 @@ void goto_convertt::convert_expression(const codet &code, goto_programt &dest)
     err_location(code);
-    std::cerr << "expression statement takes one operand\n";
-    abort();
+    throw std::runtime_error("expression statement takes one operand\n");
```
could you replace it with message.error?

mikhailramalho commented on src/goto-programs/goto_convert_functions.cpp:
```diff
@@ -103,7 +103,10 @@ void goto_convert_functionst::convert_function(symbolt &symbol)
+  auto it = functions.function_map.find(identifier);
+  if(it == functions.function_map.end())
+    functions.function_map.emplace(identifier, message_handler);
+  goto_functiont &f = functions.function_map.at(identifier);
```
why this change here?

mikhailramalho commented on src/goto-programs/goto_function.cpp:
```diff
@@ -71,9 +72,8 @@ void goto_convertt::do_function_call(
-              << std::endl;
-    abort();
+    message_handler.error(fmt::format(
+      "unexpected function argument: {}", new_function.id_string()));
```
missing abort

mikhailramalho commented on src/goto-programs/goto_program_irep.cpp:
```diff
@@ -122,9 +122,9 @@ void convert(const irept &irep, goto_programt &program)
-        throw 0;
+        throw std::runtime_error(
+          "Warning: could not resolve target link "
+          "during irep->goto_program translation.");
```
could you replace it with message.error + abort?

mikhailramalho commented on src/goto-programs/read_bin_goto_object.cpp:
```diff
@@ -89,7 +89,10 @@ bool read_bin_goto_object(
+      auto it = functions.function_map.find(symbol.id);
+      if(it == functions.function_map.end())
+        functions.function_map.emplace(symbol.id, message_handler);
+      functions.function_map.at(symbol.id).type = to_code_type(symbol.type);
```
why this change?

rafaelsamenezes replied:
Same as below

mikhailramalho commented on src/goto-programs/read_bin_goto_object.cpp:
```diff
@@ -100,7 +103,10 @@ bool read_bin_goto_object(
+    auto it = functions.function_map.find(fname);
+    if(it == functions.function_map.end())
+      functions.function_map.emplace(fname, message_handler);
+    goto_functiont &f = functions.function_map.at(fname);
```
why this change?

rafaelsamenezes replied:
Because I neeeded to pass the msg parameter to the constructor. For what I understood if you try to access an element in a map with the `[]` operator it will create the element if it doesn't exist (which caused compilation issues).

mikhailramalho commented on src/goto-symex/reachability_tree.cpp:
```diff
@@ -239,11 +239,8 @@ reachability_treet::decide_ileave_direction(execution_statet &ex_state)
-  }
-
+    message_handler.error(
+      "Ileave code selected different thread from user choice");
```
should we abort here?

mikhailramalho commented on src/goto-symex/reachability_tree.cpp:
```diff
@@ -813,11 +743,11 @@ bool reachability_treet::restore_from_dfs_state(void *)
+{
+msg.error("Unexpected number of threads when reexploring checkpoint");
+abort();
+}
```
code style?

mikhailramalho commented on src/goto-symex/reachability_tree.cpp:
```diff
@@ -829,10 +759,9 @@ bool reachability_treet::restore_from_dfs_state(void *)
-    }
+msg.error("Interleave at unexpected location when restoring checkpoint").
+abort();
+}
```
code style?

mikhailramalho commented on src/goto-symex/reachability_tree_cin.cpp:
```diff
@@ -0,0 +1,73 @@
+#include <iostream>
+
+// TODO: This is the only place with istream. How to fix this?
+int reachability_treet::get_ileave_direction_from_user() const
```
Yeah, I couldn't find any alternative either. Maybe we can move it to scanf in the future?

mikhailramalho commented on src/goto-symex/symex_function.cpp:
```diff
@@ -37,8 +37,9 @@ bool goto_symext::get_unwind_recursion(
     {
-      std::cout << "**** WARNING: k-induction does not support recursion yet. "
-                << "Disabling inductive step\n";
+      this->msg.warning(
```
no need for this->

mikhailramalho commented on src/solvers/bitwuzla/bitwuzla_conv.cpp:
```diff
@@ -1,30 +1,33 @@
-  std::cerr << "Bitwuzla error encountered\n";
-  std::cerr << msg << '\n';
-  abort();
+  assert(0 && fmt::format("Bitwuzla error encountered\n{}", msg).c_str());
```
we need an abort here, for when we build esbmc without assertions: the program should end after it printed the error message.

mikhailramalho commented on src/solvers/bitwuzla/bitwuzla_conv.cpp:
```diff
@@ -820,18 +822,24 @@ bitwuzla_convt::convert_array_of(smt_astt init_val, unsigned long domain_width)
+  FILE *f = msg.get_temp_file();
+  bitwuzla_dump_formula(bitw, "smt2", f);
+  msg.insert_and_close_file_contents(VerbosityLevel::Debug, f);
 }
```
I don't like this interface... but I don't see a better solution either.

mikhailramalho commented on src/solvers/boolector/boolector_conv.cpp:
```diff
@@ -1,35 +1,38 @@
-  std::cerr << "Boolector error encountered\n";
-  std::cerr << msg << '\n';
-  abort();
+  assert(0 && fmt::format("Boolector error encountered\n{}", msg).c_str());
```
abort?

mikhailramalho commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -2099,10 +2102,13 @@ expr2tc smt_convt::get(const expr2tc &expr)
+      {
+        std::runtime_error(
+          "Fetching array elements inside tuples currently "
+          "unimplemented, sorry");
```
missing message.error?

mikhailramalho commented on src/solvers/smt/smt_conv.cpp:
```diff
@@ -2594,61 +2602,55 @@ expr2tc smt_convt::get_by_value(const type2tc &type, BigInt value)
   abort();
 }
 
-smt_sortt smt_convt::mk_bv_sort(std::size_t width)
```
nice

mikhailramalho commented on src/solvers/smtlib/smtlib_conv.cpp:
```diff
@@ -232,10 +238,7 @@ smtlib_convt::smtlib_convt(const namespacet &_ns, const optionst &_options)
-    std::cerr << "Bad get-info :name response from solver";
-    abort();
-  }
+    std::runtime_error("Bad get-info :name response from solver");
```
missing message.error  + abort

mikhailramalho commented on src/solvers/smtlib/smtlib_conv.cpp:
```diff
@@ -257,10 +260,8 @@ smtlib_convt::smtlib_convt(const namespacet &_ns, const optionst &_options)
-    std::cerr << "Bad get-info :version response from solver";
-    abort();
-  }
+    std::runtime_error("Bad get-info :version response from solver");
```
missing message.error  + abort

mikhailramalho commented on src/solvers/smtlib/smtlib_conv.cpp:
```diff
@@ -696,14 +701,15 @@ bool smtlib_convt::get_bool(smt_astt a)
   {
-    std::cerr << "Error from smtlib solver when fetching literal value: \""
-              << smtlib_output->data << "\"" << std::endl;
+    std::runtime_error(fmt::format(
```
missing message.error

mikhailramalho commented on src/util/c_sizeof.cpp:
```diff
@@ -24,27 +24,8 @@ exprt c_sizeof(const typet &src, const namespacet &ns)
-  {
-    std::cerr << "Sizeof symbolic type encountered" << std::endl;
-    abort();
-  }
```
nice

mikhailramalho commented on src/util/context.cpp:
```diff
@@ -45,8 +46,8 @@ bool contextt::move(symbolt &symbol, symbolt *&new_symbol)
-  std::cout << std::endl << "Symbols:" << std::endl;
-
+  default_message msg;
+  msg.debug("\nSymbols:");
```
no need to create another object here, right?

mikhailramalho commented on src/util/irep2.h:
```diff
@@ -330,6 +330,10 @@ class type2t : public std::enable_shared_from_this<type2t>
   {
+    virtual const char *what() const throw()
+    {
+      return "Sizeof symbolic type encountered";
```
It should be only "symbolic type encountered"

mikhailramalho commented on src/util/irep2_type.h:
```diff
@@ -559,6 +565,14 @@ class array_type2t : public array_type_methods
+    virtual const char *what() const throw()
+    {
+      return "Sizeof nondeterministically sized array encountered";
+    }
```
remove the sizeof

mikhailramalho commented on src/util/irep_serialization.cpp:
```diff
@@ -82,8 +82,7 @@ void irep_serializationt::read_irep(std::istream &in, irept &irep)
   {
-    std::cerr << "irep not terminated. " << std::endl;
-    throw 0;
+    throw std::runtime_error("irep not terminated");
```
message.abort

mikhailramalho commented on src/util/language_file.cpp:
```diff
@@ -226,11 +225,11 @@ void language_filest::typecheck_virtual_methods(contextt &context)
               "method implementation ");
-            std::cerr << "CONVERSION ERROR" << std::endl;
+            msg.error("CONVERSION ERROR");
             throw 0;
```
abort?

mikhailramalho reviewed (approved):
Excellent work, Rafael! I've added some small comments but my main concern is that added code to the place where we search for the goto function; the message interface shouldn't break that.

Some thought/minor comments/questions:
1. Maybe we can move the abort to inside message.abort in the future? That would save some more lines of code. WDYT?
2. Do you think we can get rid of the `fmt::format`s? 
3. There is an iostream include in parseoption.cpp, esbmc still compiles if we remove it.

Overall, I'm really happy with this patch, it's something I wanted to do for a long time but I never had the courage to change all these 3 KLOC+! Please, fix the comments and it's ready to merge!

mikhailramalho commented:
BTW, ESBMC is 9MB smaller after this patch! From 255MB -> 246MB when building it with gcc 9.3.0 in RelWithDebInfo mode.

rafaelsamenezes commented:
> Maybe we can move the abort to inside message.abort in the future? That would save some more lines of code. WDYT?

I could make the default implementation as

```c
static void abort(const char* e) {
  assert(0 && msg); //  In some cases it is too hard to pass down an object just to throw an error
  abort(); // This is needed because some compilers do not detect that this is unreachable
}
```

What do you think? This should make easier to do any transition in the future.

> Do you think we can get rid of the fmt::formats?

At first I was going to implement as a variadic function (as you suggested), however not only `fmt::format` does compile-time check but it is also integrated at C++20 (which we will eventually change to). 

> There is an iostream include in parseoption.cpp, esbmc still compiles if we remove it.

Oh, I will remove it

mikhailramalho commented:
Em seg., 5 de jul. de 2021 às 08:36, Rafael Sá Menezes <
***@***.***> escreveu:

> Maybe we can move the abort to inside message.abort in the future? That
> would save some more lines of code. WDYT?
>
> I could make the default implementation as
>
> static void abort(const char* e) {
>   assert(0 && msg); //  In some cases it is too hard to pass down an object just to throw an error
>   abort(); // This is needed because some compilers do not detect that this is unreachable
> }
>
>
Actually, I was thinking of changing:

virtual void
error(const std::string &message, const std::string location = "") const
{
print(VerbosityLevel::Error, message, location);
}

to:

virtual void
error(const std::string &message, const std::string location = "") const
{
print(VerbosityLevel::Error, message, location);
abort();
}

Then removing all abort calls after every message.error call. BTW, this
discussion is for a future PR, not this one.

What do you think? This should make easier to do any transition in the
> future.
>
> Do you think we can get rid of the fmt::formats?
>
> At first I was going to implement as a variadic function (as you
> suggested), however not only fmt::format does compile-time check but it
> is also integrated at C++20 (which we will eventually change to).
>
> There is an iostream include in parseoption.cpp, esbmc still compiles if
> we remove it.
>
> Oh, I will remove it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/456#issuecomment-874079452>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6NBE66Q6JJORL7MS3TWGROHANCNFSM466JH6UQ>
> .
>


-- 

Mikhail Ramalho.

rafaelsamenezes committed:
changed throws into message error

Took 1 hour 45 minutes

rafaelsamenezes committed:
deactivated test-case

Took 6 minutes

rafaelsamenezes committed:
code sytle changes

rafaelsamenezes committed:
added missing end parenthesis

mikhailramalho commented:
@rafaelsamenezes, did you fix all my comments?

rafaelsamenezes commented:
Yes, I've implemented them
